### DISCUSSION TITLE
Using the transformer without artisan

### DISCUSSION THREAD

**Discussion by michaelhue** [spatie/typescript-transformer#26] [Ideas] [answered]
I think it would be great to be able to use the transformer without artisan, e.g. in non-laravel apps or in laravel packages that don't have artisan installed.

Maybe this could be achieved by providing a more general CLI as a [vendor binary](https://getcomposer.org/doc/articles/vendor-binaries.md)? The path to the config file could be provided as an argument.

If you are open to the idea, I'd be happy to give it a shot and create a pull request.

Thanks,
Michael

**Comment by freekmurze** [comment, marked as answer]
To keep things simple, we're going to keep a CLI outside the scope of this package.

Feel free to create a repo of your own in which you create the CLI and require the things you need from our package.

**Comment by michaelhue** [comment]
Thanks for the reply!

I currently don't have the capacity to be a maintainer for a package like this, so I went with a different route and created a custom binary in my project.

For anyone looking to do the same, here is the code:

```php
#!/usr/bin/env php
<?php

require __DIR__ . '/../vendor/autoload.php';

use Spatie\TypeScriptTransformer\Formatters\PrettierFormatter;
use Spatie\TypeScriptTransformer\Structures\TransformedType;
use Spatie\TypeScriptTransformer\TypeScriptTransformer;
use Spatie\TypeScriptTransformer\TypeScriptTransformerConfig;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\SingleCommandApplication;
use Symfony\Component\Console\Style\SymfonyStyle;

/**
 * Transformers take PHP classes(e.g., enums) as an input and will output
 * a TypeScript representation of the PHP class.
 */
$transformers = [
  Spatie\TypeScriptTransformer\Transformers\SpatieEnumTransformer::class,
  Spatie\TypeScriptTransformer\Transformers\DtoTransformer::class,
];

(new SingleCommandApplication())
  ->setName('TypeScript Transform')
  ->addArgument('output_file', InputArgument::REQUIRED, 'Writes the generated TypeScript to this file')
  ->addArgument('paths', InputArgument::REQUIRED | InputArgument::IS_ARRAY, 'Paths with classes to transform')
  ->addOption('enums', 'e', InputOption::VALUE_NONE, 'Use native TypeScript enums')
  ->addOption('format', 'f', InputOption::VALUE_NONE, 'Format the output')
  ->setCode(function (
    InputInterface $input,
    OutputInterface $output
  ) use ($transformers): int {
    $io = new SymfonyStyle($input, $output);

    $config = TypeScriptTransformerConfig::create()
      ->autoDiscoverTypes(...$input->getArgument('paths'))
      ->transformers($transformers)
      ->transformToNativeEnums($input->getOption('enums'))
      ->formatter($input->getOption('format') ? PrettierFormatter::class : null)
      ->outputFile($input->getArgument('output_file'));

    $transformer = new TypeScriptTransformer($config);

    try {
      $results = $transformer->transform();
    } catch (Exception $exception) {
      $io->error($exception->getMessage());

      return 1;
    }

    $io->table(
      ['PHP class', 'TypeScript entity'],
      array_map(
        fn (string $class, TransformedType $type) => [$class, $type->getTypeScriptName()],
        array_keys((array) $results->getIterator()),
        array_values((array) $results->getIterator())
      )
    );

    $io->info("Transformed {$results->count()} PHP types to TypeScript");

    return 0;
  })
  ->run();

```

This code goes in a `bin/typescript-transform` file in my project. 

Run this to see the available options:

```bash
./bin/typescript-transform --help
```

Which will print the following information:

```
Usage:
  ./bin/typescript-transform [options] [--] <output_file> <paths>...

Arguments:
  output_file           Writes the generated TypeScript to this file
  paths                 Paths with classes to transform

Options:
  -e, --enums           Use native TypeScript enums
  -f, --format          Use the pretty formatter
```

Use it like this:

```bash
./bin/typescript-transform --format resources/types/generated.d.ts src
```

Hope that helps!

**Reply by rubenvanassche** [reply]
Looking great @michaelhue!

**Reply by michaelhue** [reply]
Thanks @rubenvanassche! Had a lot of fun writing this (first time using the `symfony/process` package directly) and it does the job.

Let me know if you change your mind on including a general CLI in the package in the future. I'd still be happy to provide a pull request (with an added option for using a config file). Since this package already requires `symfony/process`, the added overhead would be minimal. :)

**Reply by rubenvanassche** [reply]
Hi @michaelhue, I'm currently quite busy working on the spatie/laravel-data package. Once that's done I'll take a look again at this one!

**Reply by adam-code-labx** [reply]
@rubenvanassche 
I would love to contribute and submit a PR specific for packages this would still use artisan commands programmatically
To stay within the scope of this package, I am suggesting creating a trait and contract that can be simply added to generator commands.

For example:
```php
#[AsCommand(name: 'rest-presenter:make-data')]
class MakeData extends GeneratorCommand implements TypeScriptGeneratorContract
{
    use WithTypeScriptGenerator;

    protected $name = 'rest-presenter:make-data';
    
    ....
```     

Currently I'm finalising my API package 
https://rest-presenter.com Launching soon on ProductHunt. 

I'm using Spatie data for generating and presenting data. Just finishing up the API Starter Kit for Filament, and the plan is to automate TypeScript generation for all resources.

As part of my pre-release process, I'll be developing a NuxtHub app to showcase all features, with TypeScript generation being a key focus for my presenter data layers, especially for demonstration purposes with Nuxt.

**Reply by adam-code-labx** [reply]
https://github.com/xtend-packages/rest-presenter/pull/22/files#diff-ac2b9ae1c180a560fd3bbf1b15f6c1e07ff63d735c975f8d9353c0911e53c442

@rubenvanassche 
I have this working now with my package. Let me know if this is useful and I will submit a PR
Should I submit here or perhaps in laravel-typescript-transformer repo?

![CleanShot 2024-04-06 at 17 57 35@2x](https://github.com/spatie/typescript-transformer/assets/53559175/dd986340-9261-4bf2-89fb-911aad94f23c)

**Reply by rubenvanassche** [reply]
Hi @adam-code-labx, Sorry for the late reply. Currently I'm working on v3 of TS transformer which I hope to release soon. This is a complete rewrite of the package, it even includes a module writer like you have. You're always welcome to add some extra stuff over there when it is released! At the moment I would wait with adding things to v2 since v3 hopefully will get released within a few weeks.
